Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo

```diff
@@ -0,0 +1,180 @@
+  Buildings.Controls.OBC.CDL.Interfaces.BooleanOutput triggerStart
+    "Connector for a boolean signal, true if the tuning starts"
+    annotation (Placement(
+        transformation(extent={{100,-10},{120,10}}), iconTransformation(extent={
+            {100,-10},{120,10}})));
+  Buildings.Controls.OBC.CDL.Interfaces.BooleanOutput triggerEnd
+    "Connector for a boolean signal, true if the tuning completes"
+    annotation (Placement(
+        transformation(extent={{100,-70},{120,-50}}), iconTransformation(extent=
+           {{100,-70},{120,-50}})));
+  Buildings.Controls.OBC.CDL.Continuous.Min tmin
+    "The minimum value of tOn and tOff"
+    annotation (Placement(transformation(extent={{-80,20},{-60,40}})));
+  Buildings.Controls.OBC.CDL.Continuous.Greater gretOntOff
+    "Check if tOn*tOff*min(tOn*tOff) is larger than 0"
+    annotation (Placement(transformation(extent={{0,40},{20,20}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant minLen(final k=0)
+    "Minimum value for the horizon length"
+    annotation (Placement(transformation(extent={{-40,20},{-20,40}})));
+  Buildings.Controls.OBC.CDL.Discrete.TriggeredSampler tOnSample(y_start=Buildings.Controls.OBC.CDL.Constants.eps)
+    "Sample tOn when the tuning period ends"
+    annotation (Placement(transformation(extent={{-80,60},{-60,80}})));
+  Buildings.Controls.OBC.CDL.Discrete.TriggeredSampler tOffSample(y_start=Buildings.Controls.OBC.CDL.Constants.eps)
+    "Sample tOff when the tuning period ends"
+    annotation (Placement(transformation(extent={{-80,-60},{-60,-80}})));
+  Buildings.Controls.OBC.CDL.Discrete.TriggeredSampler samAddtOntOff
+    "Sample the tmin when tmin is larger than 0"
+    annotation (Placement(transformation(extent={{40,40},{60,20}})));
+  Buildings.Controls.OBC.CDL.Continuous.Greater tIncrease
+    "Check if either tOn or tOff increases after they both becomes positive"
+    annotation (Placement(transformation(extent={{40,-40},{60,-20}})));
+  Buildings.Controls.OBC.CDL.Continuous.Min mintOntOff
+    "Find the smaller one between tOn and tOff"
+    annotation (Placement(transformation(extent={{-20,-80},{0,-60}})));
+  Buildings.Controls.OBC.CDL.Continuous.Max maxtOntOff
+    "Find the larger one between tOn and tOff"
+    annotation (Placement(transformation(extent={{-20,60},{0,80}})));
+  Buildings.Controls.OBC.CDL.Continuous.Divide halfPeriodRatioCal
+    "Calculate the half period ratio"
+    annotation (Placement(transformation(extent={{60,60},{80,80}})));
+  Buildings.Controls.OBC.CDL.Continuous.Add AddtOntOff
+    "Calculate the sum of tOn and tOff"
+    annotation (Placement(transformation(extent={{-80,-40},{-60,-20}})));
+  Buildings.Controls.OBC.CDL.Continuous.Multiply mul
+    "Detect if the tOn or tOff changes after both of them are larger than 0"
+    annotation (Placement(transformation(extent={{-40,-40},{-20,-20}})));
+  Buildings.Controls.OBC.CDL.Continuous.Greater gretmaxtOntOff
```

Comment:
the input 1 of the block is from the output of block `AddtOntOff` (the sum of the `tOn` and `tOff`). Is it your intention? I think it should be the output of `max(tOn, tOff)` if it is to implement `Check if either tOn or tOff is larger than 0`.

Reply:
Yes, it was my intention.  I did so because I want to avoid adding additional block of <code>max(tOn, tOff)</code>. When <code>tOn</code> and <code>tOff</code> are both larger than 0, if <code>sum(tOn, tOff)</code> is larger than 0, then <code>On or tOff is larger than 0</code>. I do agree this implementation is confusing, please let me know your thoughts.